File: cpp/questions/isValid.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    std::unordered_map<char, char> hm = {{')', '('}, {']', '['}, {'}', '{'}};
public:
    bool isValid(string s) {
        std::stack<char> st;
        st.push(0);
        for (const auto& c: s) {
            if (c=='(' || c=='[' || c=='{') {
                st.push(c);
            } else {
                if (st.top() == hm[c]) {
                    st.pop();
                } else {
                    return false;
                }
            }
        }
        return st.top()==0;
    }
};

File: python/questions/isValid.py
class Solution:
    def isValid(self, s: str) -> bool:
        dic = {')': '(', ']': '[', '}': '{'}
        stack = ['0']
        for c in s:
            if c in ['(', '[', '{']:
                stack.append(c)
            else:
                if dic[c] != stack.pop():
                    return False
        return len(stack) == 1
